ingestion_tools/dataset_configs/config-attribute-discovery.py

```diff
@@ -0,0 +1,213 @@
+        ):
+            current_entries[key] = recursive_dict_update(current_entries.get(key, {}), new_value)
+        # List and list situation
+        elif isinstance(new_value, list) and (
+            isinstance(current_entries.get(key), list) or current_entries.get(key) is None
+        ):
+            current_entries = recursive_dict_update_list_helper(current_entries, key, new_value)
+        # Dict and non-dict
+        elif not isinstance(new_value, dict) and isinstance(current_entries.get(key), dict):
+            # edge case: dict and a list, add the dict to the list
+            if isinstance(new_value, list):
+                new_list = new_value + [current_entries.get(key)]
+                # and then now it is list, list situation
+                current_entries = recursive_dict_update_list_helper(current_entries, key, new_list)
+            else:
+                print("dict conflict")
+                print("Key: ", key)
+                print("Current: ", current_entries.get(key))
+                print("New: ", new_value)
+        # List and non-list
+        elif not isinstance(new_value, list) and isinstance(current_entries.get(key), list):
+            # edge case: list and a dict, add the dict to the list
+            if isinstance(new_value, dict):
+                new_list = [new_value] + current_entries.get(key)
+                # and then now it is list, list situation
+                current_entries = recursive_dict_update_list_helper(current_entries, key, new_list)
+            # edge case: one-length list and a non-list, non-dict new_value = just keep the list
+            elif (
+                len(current_entries.get(key)) == 1
+                and not isinstance(new_value, list)
+                and not isinstance(new_value, dict)
+            ):
+                pass
+            else:
+                print("list conflict")
+                print("Key: ", key)
+                print("Current: ", current_entries.get(key))
+                print("New: ", new_value)
+        else:
+            raw_data_check(current_entries.get(key, None), new_value)
+            current_entries[key] = new_value
+
+    return current_entries
+
+
+def main():
+    all_files = [
+        os.path.join(directory_path, file)
+        for directory_path, _, filename in os.walk(os.path.expanduser("./"))
+        for file in filename
+    ]
+
+    unified_config = {}
+
+    # get all yaml files, except for the whitelist, and attempt to merge them
+    for file in all_files:
+        if file.endswith(".yaml") and file not in EXCLUDE_LIST:
+            with open(file, "r") as stream:
+                try:
+                    config_file = yaml.safe_load(stream)
+                except yaml.YAMLError as exc:
+                    print(exc)
+                except Exception:
+                    print("Error in file: ", file)
+                finally:
+                    stream.close()
```

Comment:
This is a really good practice. 👏  
One nice thing with Python's context manager is that it closes the file, when it exits the block. 

Here is some more information on it: https://medium.com/@orichcaspertheghost/pythons-with-statement-beyond-just-opening-and-closing-files-f9edf368a3be